pkg/gadgets/trace/dns/tracer/tracer.go

```diff
@@ -373,22 +374,26 @@ func (t *Tracer) listen(
 	}
 }
 
-func (t *Tracer) releaseLink(key string, l *link) {
+func (t *Tracer) releaseLink(netns uint64, l *link) {
 	l.perfRd.Close()
 	unix.Close(l.sockFd)
 	l.collection.Close()
-	delete(t.attachments, key)
+	delete(t.attachments, netns)
 }
 
-func (t *Tracer) Detach(key string) error {
-	if l, ok := t.attachments[key]; ok {
+func (t *Tracer) Detach(pid uint32) error {
+	netns, err := containerutils.GetNetNs(int(pid))
+	if err != nil {
+		return fmt.Errorf("getting network namespace of pid %d: %w", pid, err)
+	}
```

Comment:
Wasn't `containercollection.EventTypeRemoveContainer` occurring before the container was gone? The tests are failing because of:

```bash
    local-gadget-manager_test.go:502: Received: {{{local default test-local-gadget-dns001 } err stop tracing container "test-local-gadget-dns001": getting network namespace of pid 9356: stat /proc/9356/ns/net: no such file or directory}      }, Expected: {{{local default test-local-gadget-dns001 } debug tracer detached}      }
```

So, it seems like the container was gone when `Detach` was called. If it is confirmed, I was considering returning an `attachKey` to be used for detaching:

```golang
	attachKey, err := tracer.Attach(pid, eventCallback); 
	if err != nil {
		fmt.Printf("error attaching tracer: %s\n", err)
		return
	}
	defer tracer.Detach(attachKey)
```

WDYT?

Reply:
I tried it, but the result didn't convince me. It looks strange from the caller's point of view because the "attachment" object returned by the `tracer.Attach()` method seems to be something that should be managed by the tracer code internally. The caller should only call tracer methods: `tracer.Attach` and `tracer.Detach`. Instead of calling `tracer.Attach` and `attachment.Close()`.

 I pushed a proposal to do so. Let me know what you think about it.

Reply:
I think something like:

```golang 
  attachment := tracer.Attach(...)

...
  attachment.Close()
```

is a totally valid approach. 

It's very similar to the one used by the link package of cilium/ebpf https://pkg.go.dev/github.com/cilium/ebpf/link#Tracepoint

Anyway, I think the  approach that you proposed is fine too, so I'm fine with either solution.

Reply:
Yep, the usage also seems ok for me. But when I implemented it, I found it strange to have a map of attachments per netns within the `Tracer`, and another map in the `ConnectionToContainerCollection` with the same attachments but per pid. Don't you think? I don't know, I will have a look at the example you pointed out just for curiosity, but I will go ahead with the current approach 🙂.